Using an java8 optional as parameter can enhance readability in contr…

quentinproust opened the pull request:
Using an java8 optional as parameter can enhance readability in controllers and code quality.
And it's closer to the scala version.

In controller, we could use the optional to check if the query param is present.
```java
public Result index(Optional<String> version) {
  if (version.isPresent()) {
    // request contains version
  } else {
    // no version
  }
}
```

What do you think ?

# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [x] Have you added copyright headers to new files?
* [x] Have you checked that both Scala and Java APIs are updated?
* [x] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

# Helpful things

## Fixes

Only documentations, no fix

## Purpose

What does this PR do?

Enhance documentation about optional parameters in routes.

## Background Context

Why did you take this approach?

We're doing a coding dojo with colleagues to try play 2. We wanted to use optional parameters but didn't like to use null as default value. We wanted to use optional as we were in java but didn't find an example in the documentation. So I decided to suggest this pull request.

## References

Are there any relevant issues / PRs / mailing lists discussions?
None as far as I know.

wsargent commented:
Use of Optional in parameters is actually controversial, as it was not intended as a general purpose Maybe and is meant for returns where null is unsafe, not as a parameter input.  See https://stackoverflow.com/a/39005452/5266 for details.

richdougherty commented:
>  Use of Optional in parameters is actually controversial … See https://stackoverflow.com/a/39005452/5266 for details

I'm not sure I agree with that post. When it talks about something that's better for parameters vs return values I can just imagine easily reversing the argument around and it would make just as much sense. The *Collection wrapper* section tells us to avoid `Optional` for collection return types, when it's actually a totally valid use case. If we follow their advice and represent "the absence of a collection" as an empty collection then there are going to be bugs…

```
[error] (Play-Documentation/test:playRoutes) @766k4m3j0: Compilation error in /home/travis/build/playframework/playframework/documentation/manual/working/javaGuide/main/http/code/javaguide.http.routing.routes:33
```

This PR is currently not passing because a `list(Optional<String>)` method needs to be added to `Api.java`.

PromanSEW commented on documentation/manual/working/javaGuide/main/http/code/javaguide.http.routing.routes:
```diff
@@ -30,6 +30,8 @@ GET   /:page                controllers.Application.show(page)
 # #optional
 # The version parameter is optional. E.g. /api/list-all?version=3.0
 GET   /api/list-all         controllers.Api.list(version ?= null)
+or
```
This line should be `# or`

quentinproust replied:
You're absolutely right 👍

benmccann commented:
The better explanation I think is in the article that the Stackoverflow post links to: http://dolszewski.com/java/java-8-optional-use-cases/

quentinproust commented:
@PromanSEW I fixed the missing comment.

@richdougherty You're right. For the first syntax, the method will be `list(String)` and  for the optional syntax, it will be different as you noted. Do you thing it's necessary to add examples of the corresponding controller with both syntax ?

I understand the controversy on Optional usage for parameters. I'm not sure the arguments given in both articles apply in this case. The articles states that using arguments for parameters is overkill, that it would complicate client code and that other approach should be consider before using Optional (such as overloading).
In our case, there is no client code, it's a request that's mapped to the method. So no client code becoming clumsy, except for tests perhaps ?
And as far as I known, using overloading is not possible as we want to keep only one url.

quentinproust commented:
@richdougherty I didn't understand what you meant when you said the build failed with list(Optional<String>) needed to be added. Could you tell me where the controller is so that I can make the modification ?

gmethvin commented:
@quentinproust You changed the routes file but there is no matching method in the controller: https://github.com/quentinproust/playframework/blob/8b9d7aaaed5c899f413e203a8ffba77e7f354b1b/documentation/manual/working/javaGuide/main/http/code/javaguide/http/routing/controllers/Api.java#L10

I think it's totally fine to have some examples showing how to use `Optional`, as long as we're not forcing people to do it that way. In the context of input parameters I believe it is quite useful. We don't have to worry about the callsite being clumsy if Play takes care of constructing the `Optional` for you.

quentinproust commented:
@gmethvin Thank you for the help.

I modified `Api` controller to introduce a method `listOpt(Optional<String> version)` and changed the route file accordingly. The build is successfull again :)

gmethvin commented:
I forward-ported this to master
